File: csharp/Svix/Application.cs
// this file is @generated
#nullable enable
using Microsoft.Extensions.Logging;
using Svix.Models;

namespace Svix
{
    public class ApplicationListOptions : SvixOptionsBase
    {
        public bool? ExcludeAppsWithNoEndpoints { get; set; }
        public bool? ExcludeAppsWithDisabledEndpoints { get; set; }
        public ulong? Limit { get; set; }
        public string? Iterator { get; set; }
        public Ordering? Order { get; set; }

        public new Dictionary<string, string> QueryParams()
        {
            return SerializeParams(
                new Dictionary<string, object?>
                {
                    { "exclude_apps_with_no_endpoints", ExcludeAppsWithNoEndpoints },
                    { "exclude_apps_with_disabled_endpoints", ExcludeAppsWithDisabledEndpoints },
                    { "limit", Limit },
                    { "iterator", Iterator },
                    { "order", Order },
                }
            );
        }
    }

    public class ApplicationCreateOptions : SvixOptionsBase
    {
        public string? IdempotencyKey { get; set; }

        public new Dictionary<string, string> HeaderParams()
        {
            return SerializeParams(
                new Dictionary<string, object?> { { "idempotency-key", IdempotencyKey } }
            );
        }
    }

    public class Application(SvixClient client)
    {
        readonly SvixClient _client = client;

        /// <summary>
        /// List of all the organization's applications.
        /// </summary>
        public async Task<ListResponseApplicationOut> ListAsync(
            ApplicationListOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response =
                    await _client.SvixHttpClient.SendRequestAsync<ListResponseApplicationOut>(
                        method: HttpMethod.Get,
                        path: "/api/v1/app",
                        queryParams: options?.QueryParams(),
                        headerParams: options?.HeaderParams(),
                        cancellationToken: cancellationToken
                    );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ListAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// List of all the organization's applications.
        /// </summary>
        public ListResponseApplicationOut List(ApplicationListOptions? options = null)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ListResponseApplicationOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/app",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams()
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(List)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create a new application.
        /// </summary>
        public async Task<ApplicationOut> CreateAsync(
            ApplicationIn applicationIn,
            ApplicationCreateOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            applicationIn = applicationIn ?? throw new ArgumentNullException(nameof(applicationIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ApplicationOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/app",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: applicationIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(CreateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create a new application.
        /// </summary>
        public ApplicationOut Create(
            ApplicationIn applicationIn,
            ApplicationCreateOptions? options = null
        )
        {
            applicationIn = applicationIn ?? throw new ArgumentNullException(nameof(applicationIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ApplicationOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/app",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: applicationIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Create)} failed");

                throw;
            }
        }

        /// <summary>
        /// Gets or creates a new application.
        /// </summary>
        public async Task<ApplicationOut> GetOrCreateAsync(
            ApplicationIn applicationIn,
            ApplicationCreateOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            applicationIn = applicationIn ?? throw new ArgumentNullException(nameof(applicationIn));
            try
            {
                var queryParams = options?.QueryParams() ?? [];
                queryParams?.Add("get_if_exists", "true");

                var response = await _client.SvixHttpClient.SendRequestAsync<ApplicationOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/app",
                    queryParams: queryParams,
                    headerParams: options?.HeaderParams(),
                    content: applicationIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(GetOrCreateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Gets or creates a new application.
        /// </summary>
        public ApplicationOut GetOrCreate(
            ApplicationIn applicationIn,
            ApplicationCreateOptions? options = null
        )
        {
            applicationIn = applicationIn ?? throw new ArgumentNullException(nameof(applicationIn));
            try
            {
                var queryParams = options?.QueryParams() ?? [];
                queryParams?.Add("get_if_exists", "true");

                var response = _client.SvixHttpClient.SendRequest<ApplicationOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/app",
                    queryParams: queryParams,
                    headerParams: options?.HeaderParams(),
                    content: applicationIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(GetOrCreate)} failed");

                throw;
            }
        }

        /// <summary>
        /// Get an application.
        /// </summary>
        public async Task<ApplicationOut> GetAsync(
            string appId,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ApplicationOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } },
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(GetAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Get an application.
        /// </summary>
        public ApplicationOut Get(string appId)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ApplicationOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } }
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Get)} failed");

                throw;
            }
        }

        /// <summary>
        /// Update an application.
        /// </summary>
        public async Task<ApplicationOut> UpdateAsync(
            string appId,
            ApplicationIn applicationIn,
            CancellationToken cancellationToken = default
        )
        {
            applicationIn = applicationIn ?? throw new ArgumentNullException(nameof(applicationIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ApplicationOut>(
                    method: HttpMethod.Put,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } },
                    content: applicationIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(UpdateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Update an application.
        /// </summary>
        public ApplicationOut Update(string appId, ApplicationIn applicationIn)
        {
            applicationIn = applicationIn ?? throw new ArgumentNullException(nameof(applicationIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ApplicationOut>(
                    method: HttpMethod.Put,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } },
                    content: applicationIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Update)} failed");

                throw;
            }
        }

        /// <summary>
        /// Delete an application.
        /// </summary>
        public async Task<bool> DeleteAsync(
            string appId,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<bool>(
                    method: HttpMethod.Delete,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } },
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(DeleteAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Delete an application.
        /// </summary>
        public bool Delete(string appId)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<bool>(
                    method: HttpMethod.Delete,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } }
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Delete)} failed");

                throw;
            }
        }

        /// <summary>
        /// Partially update an application.
        /// </summary>
        public async Task<ApplicationOut> PatchAsync(
            string appId,
            ApplicationPatch applicationPatch,
            CancellationToken cancellationToken = default
        )
        {
            applicationPatch =
                applicationPatch ?? throw new ArgumentNullException(nameof(applicationPatch));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ApplicationOut>(
                    method: HttpMethod.Patch,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } },
                    content: applicationPatch,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(PatchAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Partially update an application.
        /// </summary>
        public ApplicationOut Patch(string appId, ApplicationPatch applicationPatch)
        {
            applicationPatch =
                applicationPatch ?? throw new ArgumentNullException(nameof(applicationPatch));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ApplicationOut>(
                    method: HttpMethod.Patch,
                    path: "/api/v1/app/{app_id}",
                    pathParams: new Dictionary<string, string> { { "app_id", appId } },
                    content: applicationPatch
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Patch)} failed");

                throw;
            }
        }
    }
}

File: php/src/Api/Application.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Api;

use Svix\Exception\ApiException;
use Svix\Models\ApplicationIn;
use Svix\Models\ApplicationOut;
use Svix\Models\ApplicationPatch;
use Svix\Models\ListResponseApplicationOut;
use Svix\Request\SvixHttpClient;

class Application
{
    public function __construct(
        private readonly SvixHttpClient $client,
    ) {
    }

    /**
     * List of all the organization's applications.
     *
     * @throws ApiException
     */
    public function list(
        ?ApplicationListOptions $options = null,
    ): ListResponseApplicationOut {
        $request = $this->client->newReq('GET', '/api/v1/app');
        if (null !== $options) {
            $request->setQueryParam('exclude_apps_with_no_endpoints', $options->excludeAppsWithNoEndpoints);
            $request->setQueryParam('exclude_apps_with_disabled_endpoints', $options->excludeAppsWithDisabledEndpoints);
            $request->setQueryParam('exclude_apps_with_svix_play_endpoints', $options->excludeAppsWithSvixPlayEndpoints);
            $request->setQueryParam('limit', $options->limit);
            $request->setQueryParam('iterator', $options->iterator);
            $request->setQueryParam('order', $options->order);
        }
        $res = $this->client->send($request);

        return ListResponseApplicationOut::fromJson($res);
    }

    /**
     * Create a new application.
     *
     * @throws ApiException
     */
    public function create(
        ApplicationIn $applicationIn,
        ?ApplicationCreateOptions $options = null,
    ): ApplicationOut {
        $request = $this->client->newReq('POST', '/api/v1/app');
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setBody(json_encode($applicationIn));
        $res = $this->client->send($request);

        return ApplicationOut::fromJson($res);
    }

    /**
     * Get or create a new application.
     *
     * @throws ApiException
     */
    public function getOrCreate(
        ApplicationIn $applicationIn,
        ?ApplicationCreateOptions $options = null,
    ): ApplicationOut {
        $request = $this->client->newReq('POST', '/api/v1/app');
        if (null !== $options) {
            $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        }
        $request->setQueryParam('get_if_exists', 'true');
        $request->setBody(json_encode($applicationIn));
        $res = $this->client->send($request);

        return ApplicationOut::fromJson($res);
    }

    /**
     * Get an application.
     *
     * @throws ApiException
     */
    public function get(
        string $appId,
    ): ApplicationOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}");
        $res = $this->client->send($request);

        return ApplicationOut::fromJson($res);
    }

    /**
     * Update an application.
     *
     * @throws ApiException
     */
    public function update(
        string $appId,
        ApplicationIn $applicationIn,
    ): ApplicationOut {
        $request = $this->client->newReq('PUT', "/api/v1/app/{$appId}");
        $request->setBody(json_encode($applicationIn));
        $res = $this->client->send($request);

        return ApplicationOut::fromJson($res);
    }

    /**
     * Delete an application.
     *
     * @throws ApiException
     */
    public function delete(
        string $appId,
    ): void {
        $request = $this->client->newReq('DELETE', "/api/v1/app/{$appId}");
        $res = $this->client->sendNoResponseBody($request);
    }

    /**
     * Partially update an application.
     *
     * @throws ApiException
     */
    public function patch(
        string $appId,
        ApplicationPatch $applicationPatch,
    ): ApplicationOut {
        $request = $this->client->newReq('PATCH', "/api/v1/app/{$appId}");
        $request->setBody(json_encode($applicationPatch));
        $res = $this->client->send($request);

        return ApplicationOut::fromJson($res);
    }
}
